File: src/pipeline/OutputParser.ts
// ============================================================================
// SIREEN — Output Parser
// ============================================================================
// Parses raw forge test output to extract exploit success, attacker profit,
// token balances, money flow, reverted transactions, and gas usage.

import type {
  ForgeOutput,
  ExploitResult,
  TokenBalances,
  MoneyFlowEntry,
  RevertedTx,
  GasUsage,
  ForgeTestResult,
} from './types';

export class OutputParser {
  /**
   * Parse raw forge output into structured exploit result.
   */
  parseExploitResult(
    forgeOutput: ForgeOutput,
    attackerAddress?: string
  ): ExploitResult {
    const raw = forgeOutput.raw;
    const testResults = forgeOutput.testResults;

    // Determine exploit success from test results
    const exploitTest = testResults.find(
      (t) => t.name === 'testExploit' || t.name.includes('testExploit')
    );

    const success = exploitTest?.status === 'pass';

    // Extract attacker profit from assertions and logs
    const attackerProfit = this.extractAttackerProfit(raw, success);
    const profitToken = this.extractProfitToken(raw);

    // Extract token balances
    const tokenBalances = this.extractTokenBalances(raw);

    // Extract money flow
    const moneyFlow = this.extractMoneyFlow(raw);

    // Extract reverted transactions
    const revertedTransactions = this.extractRevertedTransactions(raw);

    // Extract gas usage
    const gasUsage = this.extractGasUsage(raw, testResults);

    // Calculate approximate USD value (placeholder — no price feed in test)
    const profitUSD = 0;

    return {
      success,
      attackerProfit: attackerProfit || (success ? 'Unknown (test passed)' : 'N/A'),
      profitToken,
      profitUSD,
      tokenBalances,
      moneyFlow,
      revertedTransactions,
      gasUsage,
    };
  }

  /**
   * Extract attacker profit amount from output.
   * Looks for assert statements, balance checks, and log patterns.
   */
  private extractAttackerProfit(raw: string, exploitSuccessful: boolean): string {
    if (!exploitSuccessful) return '0';

    // Pattern: "Balance: attacker = X TOKEN"
    const balancePatterns = [
      /attacker.*?balance[:\s]*(\d+[.]?\d*)/i,
      /profit[:\s]*(\d+[.]?\d*)/i,
      /drained[:\s]*(\d+[.]?\d*)/i,
      /stolen[:\s]*(\d+[.]?\d*)/i,
      /gain[:\s]*(\d+[.]?\d*)/i,
    ];

    for (const pattern of balancePatterns) {
      const match = raw.match(pattern);
      if (match) return match[1];
    }

    // Try to extract from assertEq(attacker.balance, X) patterns
    // These are common in PoC tests
    const assertMatch = raw.match(
      /assertEq.*?attacker.*?balance.*?(\d+[.]?\d*\s*(?:ether|wei)?)/i
    );
    if (assertMatch) return assertMatch[1].trim();

    return 'See forge output for details';
  }

  /**
   * Extract the token symbol used in the profit.
   */
  private extractProfitToken(raw: string): string {
    const tokenPatterns = [
      /profit.*?token[:\s]*(\w+)/i,
      /in\s+(\w+)\s+token/i,
      /(\w+)\s+drained/i,
      /(\w+)\s+stolen/i,
    ];

    for (const pattern of tokenPatterns) {
      const match = raw.match(pattern);
      if (match) return match[1].toUpperCase();
    }

    return 'ETH';
  }

  /**
   * Extract token balances for all addresses mentioned.
   */
  private extractTokenBalances(raw: string): TokenBalances {
    const balances: TokenBalances = {};

    // Pattern: "address token balance: value"
    // e.g., "0x... ETH balance: 100.0" or "attacker DAI balance: 50000"
    const balanceRegex = /(\w+(?:\[\d+\])?|0x[a-fA-F0-9]{40})\s+(\w+)\s+balance[:\s]*(\d+[.]?\d*)/gi;

    let match: RegExpExecArray | null;
    while ((match = balanceRegex.exec(raw)) !== null) {
      const address = match[1].toLowerCase();
      const token = match[2].toUpperCase();
      const amount = match[3];

      if (!balances[address]) {
        balances[address] = {};
      }
      balances[address][token] = amount;
    }

    return balances;
  }

  /**
   * Extract money flow entries from forge output.
   * Looks for Transfer events, console.log statements, and call traces.
   */
  private extractMoneyFlow(raw: string): MoneyFlowEntry[] {
    const flow: MoneyFlowEntry[] = [];

    // Pattern: Transfer(from, to, amount) events
    const transferRegex =
      /Transfer\(?\s*(?:from:\s*)?(\w+|0x[a-fA-F0-9]{40})\s*,?\s*(?:to:\s*)?(\w+|0x[a-fA-F0-9]{40})\s*,?\s*(?:value:\s*)?(\d+[.]?\d*)\s*(?:wei|ether)?\)?/gi;

    let match: RegExpExecArray | null;
    while ((match = transferRegex.exec(raw)) !== null) {
      flow.push({
        from: match[1],
        to: match[2],
        token: 'ETH',
        amount: match[3],
        type: 'transfer',
      });
    }

    // Pattern: console.log money flow
    const logRegex = /(?:from|sender)[:\s]*(\w+|0x[a-fA-F0-9]{40})[\s,]+(?:to|receiver)[:\s]*(\w+|0x[a-fA-F0-9]{40})[\s,]+(?:amount|value)[:\s]*(\d+[.]?\d*)/gi;

    while ((match = logRegex.exec(raw)) !== null) {
      flow.push({
        from: match[1],
        to: match[2],
        token: 'ETH',
        amount: match[3],
        type: 'transfer',
      });
    }

    return flow;
  }

  /**
   * Extract reverted transactions from forge output.
   */
  private extractRevertedTransactions(raw: string): RevertedTx[] {
    const reverted: RevertedTx[] = [];

    // Pattern: [FAIL. Reason: ...] or "revert" messages
    const failRegex =
      /\[FAIL\.\s*Reason:\s*(.*?)\]\s*(\w+)?|Revert\s*(.*?)$|reverted\s*(?:with\s*)?(.*?)$/gim;

    let match: RegExpExecArray | null;
    let index = 0;
    while ((match = failRegex.exec(raw)) !== null) {
      const reason = (match[1] || match[3] || match[4] || 'Unknown').trim();
      reverted.push({
        index: index++,
        reason,
        gasUsed: 0, // Will be updated if gas data is available
      });
    }

    return reverted;
  }

  /**
   * Extract gas usage from forge output and test results.
   */
  private extractGasUsage(raw: string, testResults: ForgeTestResult[]): GasUsage {
    const byOperation: Record<string, number> = {};

    // Get gas from test results
    for (const test of testResults) {
      if (test.gasUsed) {
        byOperation[test.name] = test.gasUsed;
      }
    }

    // Parse gas report section
    const gasSection = raw.match(/Gas Report:[\s\S]*?(?=\n\n|\n─|$)/);
    if (gasSection) {
      const lines = gasSection[0].split('\n');
      for (const line of lines) {
        const fnMatch = line.match(/\|(.+?)\|.*?\|\s*(\d+)\s*\|/);
        if (fnMatch) {
          byOperation[fnMatch[1].trim()] = parseInt(fnMatch[2], 10);
        }
      }
    }

    const total = Object.values(byOperation).reduce((a, b) => a + b, 0);

    return { total, byOperation };
  }

  /**
   * Check if the forge output indicates compile errors vs test failures.
   */
  hasCompilationErrors(raw: string): boolean {
    return (
      raw.includes('Compiler run failed') ||
      raw.includes('Error: Compiler') ||
      raw.includes('ParserError:') ||
      raw.includes('TypeError:') ||
      raw.includes('DeclarationError:') ||
      (raw.includes('Error') && raw.includes('sol'))
    );
  }

  /**
   * Extract overall outcome summary from forge output.
   */
  extractOutcomeSummary(raw: string): string {
    if (this.hasCompilationErrors(raw)) {
      return 'compilation_error';
    }
    if (raw.includes('[PASS]')) {
      return raw.includes('[FAIL]') ? 'partial' : 'success';
    }
    return 'failure';
  }

  /**
   * Parse forge output into structured data including traces, state changes, and transfers.
   * Used by HonestSignal for strict verification.
   */
  parse(forgeOutput: ForgeOutput): {
    traces: any[];
    stateChanges: any[];
    transfers: any[];
  } {
    const raw = forgeOutput.raw;
    const traces: any[] = [];
    const stateChanges: any[] = [];
    const transfers: any[] = [];

    // Extract call traces from forge output (traces are shown with -vvv)
    // Pattern: "call <address>.<method>() returned <value>"
    const callTraceRegex = /\[(?:TRACE|CALL|STATICCALL|DELEGATECALL)\]\s+(?:from\s+)?(0x[a-fA-F0-9]{40}|\w+)\s+(?:to\s+)?(0x[a-fA-F0-9]{40}|\w+)\s+([^\s]+)\([^)]*\)\s+(?:returned|reverted)?\s*([^\n]*)/gi;
    
    let match: RegExpExecArray | null;
    while ((match = callTraceRegex.exec(raw)) !== null) {
      traces.push({
        from: match[1],
        to: match[2],
        method: match[3],
        result: match[4] || 'success',
        timestamp: Date.now(),
      });
    }

    // Extract storage changes (state changes)
    // Pattern: "slot <slot> changed from <old> to <new>"
    const storageChangeRegex = /slot\s+(0x[a-fA-F0-9]{64}|\d+)\s+changed\s+from\s+(\S+)\s+to\s+(\S+)/gi;
    while ((match = storageChangeRegex.exec(raw)) !== null) {
      stateChanges.push({
        slot: match[1],
        oldValue: match[2],
        newValue: match[3],
        change: match[3] !== match[2],
        timestamp: Date.now(),
      });
    }

    // Extract Transfer events from logs
    // Pattern: "Transfer(from, to, value)"
    const transferEventRegex = /Transfer\s*\(\s*(?:from:\s*)?(0x[a-fA-F0-9]{40}|\w+)\s*,\s*(?:to:\s*)?(0x[a-fA-F0-9]{40}|\w+)\s*,\s*(?:value:\s*)?(\d+)\s*\)/gi;
    while ((match = transferEventRegex.exec(raw)) !== null) {
      transfers.push({
        from: match[1],
        to: match[2],
        amount: match[3],
        token: 'ETH', // default, could be ERC20
        type: 'transfer',
        timestamp: Date.now(),
      });
    }

    // Also parse from money flow extraction (existing logic)
    const moneyFlow = this.extractMoneyFlow(raw);
    transfers.push(...moneyFlow);

    return { traces, stateChanges, transfers };
  }
}

File: gaolaif/backend/verification/output_parser.py
import re
from models.types import ForgeTestResult


class OutputParser:
    """
    Parses raw Forge output into structured, verifiable data.

    Extracts:
    - Compilation success/failure
    - Individual test results (name, pass/fail, gas, error)
    - Suite totals
    - Assertion messages
    """

    COMPILE_SUCCESS = re.compile(r"Compiler run successful|Compiler run successful", re.IGNORECASE)
    TEST_RESULT = re.compile(
        r"\[(PASS|FAIL|SKIP)"
        r"(?::\s*[^\]]+)?\]"          # match [FAIL: reason] and [FAIL] and [FAIL. Reason: ...]
        r"\s+(\S+)\s*(?:\(gas:\s*(\d+)\))?"
    )
    SUITE_RESULT = re.compile(
        r"Suite result:\s*(ok|failed|OK|FAILED)\.\s*(\d+)\s+passed;\s*(\d+)\s*failed;\s*(\d+)\s*skipped",
        re.IGNORECASE,
    )
    COMPILE_ERROR = re.compile(r"Error\s*(?:\(\d+\))?:")
    RAN_TESTS = re.compile(r"Ran (\d+) test")
    FAIL_REASON = re.compile(
        r"\[FAIL:\s*([^\]]+)\]|\[FAIL\. Reason:\s*(.+?)\]|Reason:\s*(.+?)$", re.MULTILINE
    )

    @classmethod
    def parse(cls, raw_output: str) -> dict:
        """
        Returns a structured dict:
        {
            "compiled": bool,
            "compilation_errors": list[str],
            "tests": [{"name": str, "passed": bool, "gas": int, "error": str}, ...],
            "total_passed": int,
            "total_failed": int,
            "total_skipped": int,
            "suite_ok": bool,
            "any_test_executed": bool,
        }
        """
        result = {
            "compiled": False,
            "compilation_errors": [],
            "tests": [],
            "total_passed": 0,
            "total_failed": 0,
            "total_skipped": 0,
            "suite_ok": False,
            "any_test_executed": False,
        }

        # Check compilation
        has_compile_ok = bool(cls.COMPILE_SUCCESS.search(raw_output))
        has_compile_err = bool(cls.COMPILE_ERROR.search(raw_output))
        has_compilation_ok_tag = "[COMPILATION OK]" in raw_output
        has_compilation_fail_tag = "[COMPILATION FAILED" in raw_output

        if has_compilation_fail_tag:
            result["compiled"] = False
            result["compilation_errors"] = cls._extract_compile_errors(raw_output)
            return result

        if has_compilation_ok_tag or has_compile_ok:
            result["compiled"] = True
        elif has_compile_err:
            result["compiled"] = False
            result["compilation_errors"] = cls._extract_compile_errors(raw_output)
            return result

        # Extract individual test results
        for match in cls.TEST_RESULT.finditer(raw_output):
            status = match.group(1)
            test_name = match.group(2)
            gas_str = match.group(3)
            gas = int(gas_str) if gas_str else 0

            passed = status == "PASS"
            if status == "SKIP":
                passed = False

            error_msg = ""
            if status == "FAIL":
                error_msg = cls._extract_fail_reason(raw_output, test_name)

            result["tests"].append({
                "name": test_name,
                "passed": passed,
                "gas": gas,
                "error": error_msg,
            })

            if passed:
                result["total_passed"] += 1
            elif status == "FAIL":
                result["total_failed"] += 1
            elif status == "SKIP":
                result["total_skipped"] += 1

        # Parse suite summary
        suite_match = cls.SUITE_RESULT.search(raw_output)
        if suite_match:
            result["suite_ok"] = suite_match.group(1).lower() == "ok"
            result["total_passed"] = int(suite_match.group(2))
            result["total_failed"] = int(suite_match.group(3))
            result["total_skipped"] = int(suite_match.group(4))
        elif cls.RAN_TESTS.search(raw_output):
            result["suite_ok"] = result["total_failed"] == 0

        result["any_test_executed"] = (
            result["total_passed"] > 0 or result["total_failed"] > 0
        )

        result["_raw_forge_output"] = raw_output

        return result

    @classmethod
    def parse_tests(cls, raw_output: str) -> list[ForgeTestResult]:
        tests = []
        for match in cls.TEST_RESULT.finditer(raw_output):
            status = match.group(1)
            test_name = match.group(2)
            gas_str = match.group(3)
            gas = int(gas_str) if gas_str else 0
            tests.append(ForgeTestResult(
                test_name=test_name,
                passed=(status == "PASS"),
                gas_used=gas,
                error_message=cls._extract_fail_reason(raw_output, test_name) if status == "FAIL" else "",
            ))
        return tests

    @classmethod
    def _extract_compile_errors(cls, output: str) -> list[str]:
        errors = []
        for line in output.split("\n"):
            stripped = line.strip()
            if not stripped:
                continue
            if re.match(r'^Error\b', stripped) or 'error[' in stripped.lower():
                errors.append(stripped)
            elif 'undeclared identifier' in stripped.lower():
                errors.append(stripped)
            elif errors and stripped.startswith('^'):
                errors[-1] += ' ' + stripped
        return errors[:10]

    @classmethod
    def _extract_fail_reason(cls, output: str, test_name: str) -> str:
        for line in output.splitlines():
            if "[FAIL" in line and test_name in line:
                for match in cls.FAIL_REASON.finditer(line):
                    reason = (match.group(1) or match.group(2) or match.group(3) or "").strip()
                    if reason:
                        return reason
        for match in cls.FAIL_REASON.finditer(output):
            reason = (match.group(1) or match.group(2) or match.group(3) or "").strip()
            if reason:
                return reason
        return "Assertion failed"
